[Clang][ARM] Ensure FPU Features are parsed when targeting `cc1as`

Stylie777 opened the pull request:
Previously, `cc1as` did not consider the Features that can be included from a target's FPU. This could lead to a situation where assembly files could not compile as cc1as did not know if a feature was supported.

With this change, all the features for the FPU will be passed to `cc1as` as `-target-feature` lines. By making this change, it will enable `+nosimd` to be functional, worked on in #130623, and fix a regression introduced in 8fa0f0efce5fb81eb422e6d7eec74c66dafef4a3 so armv7s-apple0darwin targets can utilise VFPv4 correctly.

Co-authored-by: Martin Storsjö <martin@martin.st>

Stylie777 committed:
[Clang] [ARM] Ensure FPU Features are collected when using the Clang Assembler

Previously, FPU features were not collected when forming a list
of features for the Assembler.

This fixes a regression from 8fa0f0efce5fb81eb422e6d7eec74c66dafef4a3,
which caused VFPv4 to be unavailable if assembling for an
armv7s-apple-darwin target.

Co-authored-by: Martin Storsjö <martin@martin.st>

Stylie777 committed:
Respond to review comments originally from #134366

These changes were originally opened under #134366 but were
moved to this PR to allow me to drive further changes as I was
the original author.

This commit responds to comments made on the original PR.

DavidSpickett commented on clang/lib/Driver/ToolChains/Arch/ARM.cpp:
```diff
@@ -679,20 +679,17 @@ llvm::ARM::FPUKind arm::getARMTargetFeatures(const Driver &D,
+      llvm::ARM::ArchKind ArchKind =
+          arm::getLLVMArchKindForARM(CPU, ArchName, Triple);
+      FPUKind = llvm::ARM::getDefaultFPU(CPU, ArchKind);
+      (void)llvm::ARM::getFPUFeatures(FPUKind, Features);
```
Could move this line after the if, given that it always gets called, it's just the value of FPUKind that changes between branches.

DavidSpickett commented on clang/test/Driver/arm-fpu-selection.s:
```diff
@@ -0,0 +1,32 @@
+// REQUIRES: arm-registered-target
+// Ensures that when targeting an ARM target with an Asm file, clang
```
I have often been asked to use `///` for comments that were actually comments. I think it is marginally easier to read if you do that.

DavidSpickett commented on clang/test/Driver/arm-fpu-selection.s:
```diff
@@ -0,0 +1,32 @@
+// activation of NEON for supported targets. The Cortex-R52 will be
+// used and tested for VFP and NEON Support
+
+// RUN: %clang -target arm-none-eabi -mcpu=cortex-r52 -c %s -o /dev/null | count 0
```
The purpose of this line is what exactly?

If we have errors the RUN line will fail because clang returns non-zero, so perhaps a check for warnings? Should you be piping stderr as well if that's the case?

DavidSpickett commented on clang/test/Driver/arm-fpu-selection.s:
```diff
@@ -0,0 +1,32 @@
+// used and tested for VFP and NEON Support
+
+// RUN: %clang -target arm-none-eabi -mcpu=cortex-r52 -c %s -o /dev/null | count 0
+// RUN: %clang -target arm-none-eabi -mcpu=cortex-r52 -c %s -o /dev/null -### 2>&1 | FileCheck --check-prefix=CHECK-TARGET-FEATURES %s
```
You only have one CHECK name (whatever the term is?) here so you could just make the check lines `CHECK:` and not need to set a check-prefix. Applies to the other files too.

Stylie777 committed:
Respond to review comments

Summary of changes
- Adds check lines to ensure there are no errors/warnings
- Makes all comments on test files `///`
- Move the `getFPUFeatures` line so it is called outside of the
if block that selects the FPU.

Stylie777 reviewed (commented):
Thanks for the review @DavidSpickett, I have made some updates based on this.

DavidSpickett commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -463,6 +466,7 @@ X86 Support
 
 Arm and AArch64 Support
 ^^^^^^^^^^^^^^^^^^^^^^^
+- For ARM targets, cc1as now considers the FPU's features for the selected CPU or Arch.
```
Arch -> architecture

Stylie777 replied:
Done

DavidSpickett reviewed (approved):
LGTM

Stylie777 committed:
Fix spelling in Release Notes

mstorsjo commented on clang/lib/Driver/ToolChains/Arch/ARM.cpp:
```diff
@@ -679,21 +679,17 @@ llvm::ARM::FPUKind arm::getARMTargetFeatures(const Driver &D,
-    }
+    std::string CPU = arm::getARMTargetCPU(CPUName, ArchName, Triple);
+    bool Generic = CPU == "generic";
     if (Generic && (Triple.isOSWindows() || Triple.isOSDarwin()) &&
         getARMSubArchVersionNumber(Triple) >= 7) {
       FPUKind = llvm::ARM::parseFPU("neon");
```
Those are somewhat still needed in order to keep things working for that patched case, yes.

It's indeed a bit odd to avoid breakage for an out-of-tree downstream patch that isn't present here, this was discussed in #122095, but there @DavidSpickett agreed that it seems good to be explicit about these things - just like for the Android case just above here.

The problem with adjusting the downstream patch is that it is for Debian/Ubuntu packaging, adjusting what is included in the armv7 baseline (for the purposes of Linux targets), but affecting how useful it is as a cross compiler for other targets. But for them to take extra effort into their Linux-targeting patch to accommodate for using it for cross compilation for other OSes feels like a big thing to request anyway (even though it is their patch which is breaking things).

> getARMCPUForArch already has some cpu overrides for different OS's.

Yeah, that would be possible as well. Windows does have one, and thanks to this patch it gets applied on both assembly and C. Darwin doesn't, but it should be possible to pick e.g. a `cortex-a8` as baseline there.

mstorsjo replied:
Yeah that's definitely the intent of it - but making it work that way definitely is less obvious than one would want...

Stylie777 commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -38,6 +38,9 @@ Potentially Breaking Changes
 - Fix missing diagnostics for uses of declarations when performing typename access,
   such as when performing member access on a '[[deprecated]]' type alias.
   (#GH58547)
+- For ARM targets, when using cc1as, the features included in the selected CPU or
+  Architecture's FPU are now loaded and utilized. If you wish not to use a specific feature,
+  the relevant ``+no`` option will need to be amended to the command line option.
```
Updated.

Stylie777 committed:
Improve release notes based on review comments.

davemgreen reviewed (approved):
The other thing this patch does is bring the preprocessor features for -march=armv8.1-m in line with the assembly features. It was previously enabling __ARM_VFPV2__ but not allowing vfpv2 instructions.

I'm still not sure about the Win/Darwin part, but otherwise LGTM.

mstorsjo commented:
> I'm still not sure about the Win/Darwin part, but otherwise LGTM.

We could fix it by adding a case for armv7 on Darwin in `ARM::getARMCPUForArch`, which would allow getting rid of that special case here. But as we currently have the special cases for Windows/Darwin here, I would rather prefer to not touch that within this change, but do such cleanup in a later separate PR.

Stylie777 commented:
> But as we currently have the special cases for Windows/Darwin here, I would rather prefer to not touch that within this change, but do such cleanup in a later separate PR.
+1 to this, I agree this should be done in a separate patch.

Thanks all for your reviews.
